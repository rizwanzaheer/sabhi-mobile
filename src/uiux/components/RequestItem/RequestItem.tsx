import React, { useState, useEffect } from 'react'
import Container from '../Container/Container'
import Icon from '../Icon/Icon'
import Text, { TextTypes } from '../Text/Text'
import Radio from '../RadioBtn/RadioBtn'
import InlineCredentialInput from '../InlineCredential/InlineCredential'
import Credential from '../Credential/Credential'
import { TouchableHighlight } from 'react-native'
import { withTheme } from '../../theming'
import { RequestItemSelectable } from '../../types'
import * as Kancha from '../../types'

const S = require('sugar/string')

interface Issuers {
  did: {
    did: string
    url?: string
  }
}

interface CredentialField {
  type: string
  value: string
}

interface RequestCredential extends Kancha.VerifiableCredential {
  fields: CredentialField[]
}

interface RequestItem {
  /**
   *  Pass a test ID to button
   */
  testID?: string
  /**
   * Remove bottom divider
   */
  last?: boolean

  /**
   * Remove bottom divider
   */
  dividerBottom?: boolean

  /**
   * Subtitle to show over main text
   */
  claimType: string

  /**
   *  Note to show on the right
   */
  itemNote?: string

  /**
   *  Credential property options
   */
  credentials: RequestCredential[]

  /**
   *  Issuers
   */
  issuers?: Issuers[]

  /**
   *  Reason
   */
  reason?: string

  /**
   *  The item being request is required
   */
  required?: boolean

  /**
   *  Return the selected item
   */
  onSelectItem: (id: string | null, jwt: string | null, claimType: string) => void

  /**
   *  When a VC is pressed
   */
  onPressVC?: (vc: Kancha.VerifiableCredential) => void
  /**
   *  Close after select item
   */
  closeAfterSelect?: boolean

  /**
   *  Function to self sign a credential
   */
  selfSign?: (claimType: string, value: string) => void

  /**
   *  The reference for the credential input
   */
  inputRef?: any

  /**
   *  Theme
   */
  theme: any
}
/**
 * Request list item for request flows
 */
const RequestItem: React.FC<RequestItem> = ({
  testID,
  required,
  credentials,
  claimType,
  onSelectItem,
  closeAfterSelect,
  onPressVC,
  last,
  dividerBottom,
  itemNote,
  reason,
  theme,
  issuers,
  inputRef,
  selfSign,
}) => {
  const [options, updateSelected] = useState<RequestItemSelectable[]>([])
  const [optionsExpanded, toggleOptions] = useState(false)
  const [shared, toggleShared] = useState(required)
  const selected = options.find((i: RequestItemSelectable) => i.selected)

  const makeSelection = (id: string | null, jwt: string | null, decline: boolean) => {
    const updatedOptions = options.map((item: any) => {
      return !decline && item.id === id ? { ...item, selected: true } : { ...item, selected: false }
    })
    updateSelected(updatedOptions)
    toggleShared(!decline)
    if (closeAfterSelect !== false) {
      toggleOptions(false)
    }
    onSelectItem(id, jwt, claimType)
  }

  useEffect(() => {
    let defaultOptions = credentials.map(
      (vc: any, index: number): RequestItemSelectable => {
        const field = vc.fields.find((field: CredentialField) => {
          return field.type === claimType
        })
        const otherFields = vc.fields
          .map((f: CredentialField) => f.type !== claimType && f.type)
          .filter((f: string | null) => f)
          .map((f: string) => S.String.titleize(f))

        return {
          id: vc.hash + '-' + index,
          jwt: vc.jwt,
          iss: vc.iss,
          ...field,
          otherFields,
          otherfieldCount: otherFields.length,
          selected: required && index === 0,
          vc,
        }
      },
    )

    updateSelected(defaultOptions)
  }, [credentials])

  return (
    <Container>
      <Container background={'primary'} dividerBottom={!last || dividerBottom}>
        <TouchableHighlight
          underlayColor={'transparent'}
          onPress={() => toggleOptions(!optionsExpanded)}
          testID={testID}
        >
          <Container
            paddingLeft={24}
            paddingBottom
            paddingTop={16}
            flexDirection={'row'}
            alignItems={'center'}
          >
            <Container flex={1}>
              {claimType && (
                <Text type={TextTypes.SubTitle}>
                  {S.String.titleize(claimType)}
                  {required ? '*' : ''}
                </Text>
              )}
              <Container paddingTop={2}>
                <Text warn={required && !selected} type={TextTypes.Body}>
                  {selected ? selected.value : 'Not Shared'}
                </Text>
              </Container>
            </Container>
            {itemNote && (
              <Container paddingRight={32}>
                <Text type={TextTypes.ListItemNote}>{itemNote}</Text>
              </Container>
            )}
            {
              <Container paddingRight={32}>
                <Icon icon={optionsExpanded ? theme.icons.UP : theme.icons.DOWN} />
              </Container>
            }
          </Container>
        </TouchableHighlight>
        {optionsExpanded && (
          <Container marginLeft={8} paddingLeft={true} paddingBottom={true} paddingRight={true}>
            <Container paddingBottom>
              <Text type={TextTypes.SubTitle}>{reason}</Text>
            </Container>
            {required && !selected && issuers && (
              <Container padding backgroundColor={'#FAD2D8'} br={5}>
                <Text type={TextTypes.Body}>
                  A {claimType} credential issued by <Text bold>{issuers[0].did.did}</Text> is required
                </Text>
              </Container>
            )}

            {!required && (
              <Container flexDirection={'row'} paddingBottom={10}>
                <Container flex={1}>
                  <Radio
                    disabled={required}
                    selected={!shared}
                    onPress={() => makeSelection(null, null, true)}
                  >
                    Do not share
                  </Radio>
                </Container>
              </Container>
            )}
            {options.map((item: RequestItemSelectable, index: number) => {
              return (
                <Container key={item.id + index}>
                  <Container
                    flexDirection={'row'}
                    alignItems={'flex-start'}
                    justifyContent={'center'}
                    paddingBottom={10}
                  >
                    <Container flex={1}>
                      <Radio selected={item.selected} onPress={() => makeSelection(item.id, item.jwt, false)}>
                        {item.value}
                      </Radio>
                    </Container>
                  </Container>
                  <Credential
                    onPress={() => onPressVC && onPressVC(item.vc)}
                    issuer={item.vc.iss}
                    subject={item.vc.sub}
                    exp={item.vc.exp}
                    fields={item.vc.fields}
                    background={item.selected ? 'primary' : 'secondary'}
                    shadow={item.selected ? 2 : 0}
                  />
                </Container>
              )
            })}
            {!issuers && selfSign && (
              <InlineCredentialInput
                inputRef={inputRef}
                claimType={claimType}
                onCreate={(value: string) => selfSign(claimType, value)}
              />
            )}
          </Container>
        )}
      </Container>
    </Container>
  )
}
export default withTheme(RequestItem)
