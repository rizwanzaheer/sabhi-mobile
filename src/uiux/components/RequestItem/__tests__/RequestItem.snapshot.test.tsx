import * as React from 'react'
import { render } from 'react-native-testing-library'
import RequestItem from '../RequestItem'

jest.mock('react-native-vector-icons/Ionicons', () => 'Icon')

const baseVc = {
  jwt: '001',
  iss: { shortId: 'Serto Issuer', did: '0xfksksdkeprgj' },
  sub: { shortId: 'Sert User', did: '0xwrhfowiehrf' },
  type: '',
  iat: 123455678,
  exp: 123455678,
  nbf: 123455678,
}

const singleOptionCredential_01 = {
  ...baseVc,
  hash: '001',
  parentHash: '00001',
  fields: [
    {
      type: 'name',
      value: 'Single Option A',
      isObj: false,
    },
  ],
}

const singleOptionCredential_02 = {
  ...baseVc,
  hash: '002',
  parentHash: '00002',
  fields: [
    {
      type: 'name',
      value: 'Single Option B',
      isObj: false,
    },
  ],
}

describe('Component(snapshot): RequestItem', () => {
  it('should render required with options', () => {
    const onSelect = jest.fn()
    const tree = render(
      <RequestItem
        required
        claimType={'name'}
        reason={'A test reason'}
        credentials={[singleOptionCredential_01, singleOptionCredential_02]}
        onSelectItem={onSelect}
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render un-required with single option', () => {
    const onSelect = jest.fn()
    const tree = render(
      <RequestItem
        claimType={'name'}
        reason={'A test reason'}
        credentials={[singleOptionCredential_01]}
        onSelectItem={onSelect}
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render un-required with multi options', () => {
    const onSelect = jest.fn()
    const selfSign = jest.fn()
    const tree = render(
      <RequestItem
        selfSign={selfSign}
        claimType={'name'}
        reason={'A test reason'}
        credentials={[singleOptionCredential_01, singleOptionCredential_02]}
        onSelectItem={onSelect}
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
