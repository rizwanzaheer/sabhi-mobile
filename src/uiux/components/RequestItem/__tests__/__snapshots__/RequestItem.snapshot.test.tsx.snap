// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(snapshot): RequestItem should render required with options 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": "#FFFFFF",
        "borderBottomColor": "#CCCCCC",
        "borderBottomWidth": 0.5,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": undefined,
            "flexDirection": "row",
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": 15,
            "paddingHorizontal": undefined,
            "paddingLeft": 24,
            "paddingRight": undefined,
            "paddingTop": 16,
            "width": undefined,
          }
        }
      >
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": 1,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#BBBBBB",
                "flexWrap": "wrap",
                "fontSize": 14,
              }
            }
          >
            Name
            *
          </Text>
          <View
            pointerEvents="auto"
            style={
              Object {
                "alignItems": undefined,
                "backgroundColor": undefined,
                "borderRadius": undefined,
                "flex": undefined,
                "flexDirection": undefined,
                "height": undefined,
                "justifyContent": undefined,
                "margin": undefined,
                "marginBottom": undefined,
                "marginLeft": undefined,
                "marginRight": undefined,
                "marginTop": undefined,
                "padding": undefined,
                "paddingBottom": undefined,
                "paddingHorizontal": undefined,
                "paddingLeft": undefined,
                "paddingRight": undefined,
                "paddingTop": 2,
                "width": undefined,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#333333",
                  "flexWrap": "wrap",
                  "fontSize": 16,
                  "lineHeight": 22,
                }
              }
            >
              Single Option A
            </Text>
          </View>
        </View>
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": undefined,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": 32,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Icon
            color="#BBBBBB"
            name="ios-arrow-down"
            size={26}
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Component(snapshot): RequestItem should render un-required with multi options 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": "#FFFFFF",
        "borderBottomColor": "#CCCCCC",
        "borderBottomWidth": 0.5,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": undefined,
            "flexDirection": "row",
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": 15,
            "paddingHorizontal": undefined,
            "paddingLeft": 24,
            "paddingRight": undefined,
            "paddingTop": 16,
            "width": undefined,
          }
        }
      >
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": 1,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#BBBBBB",
                "flexWrap": "wrap",
                "fontSize": 14,
              }
            }
          >
            Name
            
          </Text>
          <View
            pointerEvents="auto"
            style={
              Object {
                "alignItems": undefined,
                "backgroundColor": undefined,
                "borderRadius": undefined,
                "flex": undefined,
                "flexDirection": undefined,
                "height": undefined,
                "justifyContent": undefined,
                "margin": undefined,
                "marginBottom": undefined,
                "marginLeft": undefined,
                "marginRight": undefined,
                "marginTop": undefined,
                "padding": undefined,
                "paddingBottom": undefined,
                "paddingHorizontal": undefined,
                "paddingLeft": undefined,
                "paddingRight": undefined,
                "paddingTop": 2,
                "width": undefined,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#333333",
                  "flexWrap": "wrap",
                  "fontSize": 16,
                  "lineHeight": 22,
                }
              }
            >
              Not Shared
            </Text>
          </View>
        </View>
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": undefined,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": 32,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Icon
            color="#BBBBBB"
            name="ios-arrow-down"
            size={26}
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`Component(snapshot): RequestItem should render un-required with single option 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": "#FFFFFF",
        "borderBottomColor": "#CCCCCC",
        "borderBottomWidth": 0.5,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": undefined,
            "flexDirection": "row",
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": 15,
            "paddingHorizontal": undefined,
            "paddingLeft": 24,
            "paddingRight": undefined,
            "paddingTop": 16,
            "width": undefined,
          }
        }
      >
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": 1,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#BBBBBB",
                "flexWrap": "wrap",
                "fontSize": 14,
              }
            }
          >
            Name
            
          </Text>
          <View
            pointerEvents="auto"
            style={
              Object {
                "alignItems": undefined,
                "backgroundColor": undefined,
                "borderRadius": undefined,
                "flex": undefined,
                "flexDirection": undefined,
                "height": undefined,
                "justifyContent": undefined,
                "margin": undefined,
                "marginBottom": undefined,
                "marginLeft": undefined,
                "marginRight": undefined,
                "marginTop": undefined,
                "padding": undefined,
                "paddingBottom": undefined,
                "paddingHorizontal": undefined,
                "paddingLeft": undefined,
                "paddingRight": undefined,
                "paddingTop": 2,
                "width": undefined,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#333333",
                  "flexWrap": "wrap",
                  "fontSize": 16,
                  "lineHeight": 22,
                }
              }
            >
              Not Shared
            </Text>
          </View>
        </View>
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": undefined,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": undefined,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": 32,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Icon
            color="#BBBBBB"
            name="ios-arrow-down"
            size={26}
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
